File: framework/dotnet/BetterPortal.Tool/Scaffold.cs
using System.Reflection;
using System.Text;
using BetterPortal;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal.Tool;

internal static class Scaffold
{
    public static int Command(string[] args)
    {
        if (args.Length == 0) throw new ArgumentException("Use init DIRECTORY --plugin-id ID --registry-ref NAMESPACE/NAME --title TITLE");
        var options = new Dictionary<string, string>();
        for (var index = 1; index < args.Length; index++)
        {
            if (args[index] is not ("--plugin-id" or "--registry-ref" or "--title") || index + 1 >= args.Length)
                throw new ArgumentException("Unknown or incomplete argument: " + args[index]);
            options.Add(args[index], args[++index]);
        }
        if (!options.TryGetValue("--plugin-id", out var pluginId) || !options.TryGetValue("--registry-ref", out var registryRef) || !options.TryGetValue("--title", out var title))
            throw new ArgumentException("Use init DIRECTORY --plugin-id ID --registry-ref NAMESPACE/NAME --title TITLE");
        Contracts.Parse("ManifestDeclarationSchema", new Node { ["pluginId"] = pluginId, ["title"] = title, ["description"] = title, ["version"] = "1.0.0" });
        var config = Contracts.Parse("BetterPortalProjectConfigSchema", new Node { ["registryRef"] = registryRef, ["defaultNamespace"] = registryRef.Split('/')[0] });
        var runtime = typeof(Scaffold).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()!.InformationalVersion.Split('+')[0];
        var files = new Dictionary<string, string>
        {
            ["betterportal.json"] = Json.Write(config),
            ["betterportal.lock.json"] = Json.Write(Contracts.Parse("BetterPortalLockSchema", new Node())),
            [".gitignore"] = "bin/\nobj/\n.bp-state/\n.bp-generated/\n.env",
            ["BpService.csproj"] = $$"""
                <Project Sdk="Microsoft.NET.Sdk.Web">
                  <PropertyGroup>
                    <TargetFramework>net10.0</TargetFramework>
                    <Nullable>enable</Nullable>
                    <ImplicitUsings>enable</ImplicitUsings>
                    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
                    <RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
                  </PropertyGroup>
                  <ItemGroup>
                    <PackageReference Include="BetterPortal.AspNetCore" Version="[{{runtime}}]" />
                    <Compile Include="**/.well-known/**/*.cs" />
                  </ItemGroup>
                </Project>
                """,
            ["bp-routes/hello/index.cs"] = """
                using BetterPortal;
                using BetterPortal.Generated;

                namespace BpService.Routes;
                public static class Index
                {
                    [RouteModule]
                    public static RouteDeclarationInput Create() => new() { ViewId = "hello.index", Title = "Hello" };
                }
                """,
            ["bp-routes/hello/GET.cs"] = """
                using AnyVali;
                using BetterPortal;
                using BetterPortal.Generated;

                namespace BpService.Routes;
                public static class Hello
                {
                    [RouteModule]
                    public static Operation Create()
                    {
                        var page = new Renderer<string>(new() { Renderer = "bootstrap5" }, (value, context) => "<p>Hello</p>");
                        var handler = new Handler<object?, object?, object?, object?, string>(
                            V.String(), context => ValueTask.FromResult("Hello"), renderers: [page]);
                        return new(handler, new() { OperationId = "hello.get", Method = HttpMethodInput.GET,
                            Title = "Hello", Description = "Return a greeting", Auth = new() { Required = true } });
                    }
                }
                """,
            ["Definition.cs"] = $$"""
                using BetterPortal;
                using BetterPortal.Generated;

                namespace BpService;
                public static class Definition
                {
                    public static ManifestDeclarationInput Declaration() => new() {
                        PluginId = {{Json.Write(pluginId)}}, Title = {{Json.Write(title)}},
                        Description = {{Json.Write(title)}}, Version = "1.0.0"
                    };
                    public static Registry Registry() => Discovery.Discover(typeof(Definition).Assembly);
                    public static BpSchemaOutput Export() => Registry().Schema(Declaration());
                }
                """,
            ["Program.cs"] = """
                using BetterPortal;
                using BetterPortal.AspNetCore;
                using BpService;

                if (args.SequenceEqual(["--new-bootstrap-key"]))
                {
                    Console.WriteLine(BootstrapCipher.GenerateKey());
                    return;
                }
                static string Required(string name) => Environment.GetEnvironmentVariable(name)
                    ?? throw new InvalidOperationException("Configure " + name);
                var builder = WebApplication.CreateBuilder(args);
                var directory = Environment.GetEnvironmentVariable("BP_STATE_DIRECTORY")
                    ?? Path.Combine(builder.Environment.ContentRootPath, ".bp-state");
                var service = new Service(Definition.Registry(), Definition.Declaration(), managed: true,
                    stateStore: new FileStateStore(Path.Combine(directory, "snapshot.json")));
                var installation = new ServiceInstallation(service,
                    new BootstrapStateStore(new FileStateStore(Path.Combine(directory, "bootstrap.json")), Required("BP_BOOTSTRAP_MASTER_KEY")),
                    Required("BP_CP_URL"), Required("BP_PUBLIC_ORIGIN"),
                    settingsStore: new FileStateStore(Path.Combine(directory, "settings.json")));
                builder.Services.AddBetterPortal(service, installation: installation);
                var app = builder.Build();
                app.MapBetterPortal(service, installation: installation);
                await app.RunAsync();
                """,
            ["README.md"] = """
                # Standalone BP service

                Install the matching native BP packages with .NET 10, then run:

                    dotnet build
                    bp-dotnet export --assembly bin/Debug/net10.0/BpService.dll --factory BpService.Definition:Export --output bp-contract.json
                    bp-dotnet deps sync --frozen --check

                For unpublished packages, first restore with --source /path/to/local/packages
                and --source https://api.nuget.org/v3/index.json. Commit packages.lock.json and
                use dotnet restore --locked-mode for subsequent builds.
                No Node or BSB installation is needed by this service.

                Configure BP_CP_URL (trusted config manager), BP_PUBLIC_ORIGIN (this service's
                public origin), and BP_BOOTSTRAP_MASTER_KEY through your host's protected secrets.
                Generate a master key once and retain it across restarts:

                    dotnet run -- --new-bootstrap-key

                BP_STATE_DIRECTORY defaults to .bp-state under the content root; mount persistent
                storage there in a container. Keep the master key separate from that directory.
                URLs require HTTPS except the runtime's exact loopback development exceptions.

                    dotnet run -- --urls http://127.0.0.1:8000

                Install through the existing config manager using the configured public origin.
                Health at /.well-known/bp/health returns 503 until installation, manifest submission
                and a valid scoped snapshot succeed. Mount hello.index / hello.get in an app;
                /hello then returns JSON or Bootstrap HTML with valid user authentication.
                The ASP.NET hosted service owns installation, sync and shutdown cancellation.

                Edit Definition.cs for BP identity/version and bp-routes for operations. Keep
                betterportal.json and betterportal.lock.json in source control. Use native deps
                commands for dependency contracts and repeat export after changing the registry.
                """
        };
        var directory = Path.GetFullPath(args[0]);
        if (Directory.Exists(directory) || File.Exists(directory)) throw new IOException("The output directory must not exist");
        Directory.CreateDirectory(directory);
        foreach (var (name, content) in files)
        {
            var target = Path.Combine(directory, name);
            Directory.CreateDirectory(Path.GetDirectoryName(target)!);
            using var output = new StreamWriter(new FileStream(target, FileMode.CreateNew, FileAccess.Write), new UTF8Encoding(false));
            output.Write(content + "\n");
        }
        Console.WriteLine(Path.Combine(directory, "README.md"));
        return 0;
    }
}

File: framework/python/betterportal/scaffold.py
"""Create a standalone service using the installed native runtime."""
from importlib.metadata import version
import json
from pathlib import Path

from .contracts import parse


def scaffold(directory: Path, plugin_id: str, registry_ref: str, title: str) -> None:
    declaration = {"pluginId": plugin_id, "title": title, "description": title, "version": "1.0.0"}
    parse("ManifestDeclarationSchema", declaration)
    config = parse("BetterPortalProjectConfigSchema", {"registryRef": registry_ref, "defaultNamespace": registry_ref.split("/")[0]})
    runtime = version("betterportal")
    files = {
        "betterportal.json": json.dumps(config, indent=2) + "\n",
        "betterportal.lock.json": json.dumps(parse("BetterPortalLockSchema", {}), indent=2) + "\n",
        ".gitignore": ".venv/\n__pycache__/\n*.egg-info/\n.bp-state/\n.bp-generated/\n.env\n",
        "pyproject.toml": f'''[build-system]
requires = ["setuptools>=77"]
build-backend = "setuptools.build_meta"

[project]
name = {json.dumps(registry_ref.replace('/', '-'))}
version = "1.0.0"
requires-python = ">=3.10"
dependencies = ["betterportal[asgi]=={runtime}", "uvicorn==0.52.4"]

[tool.setuptools.packages.find]
include = ["my_service*"]
''',
        "my_service/__init__.py": "",
        "my_service/bp_routes/__init__.py": "",
        "my_service/bp_routes/hello/index.py": '''from betterportal.generated_types import RouteDeclarationInput

def create() -> RouteDeclarationInput:
    return {"viewId": "hello.index", "title": "Hello"}
''',
        "my_service/bp_routes/hello/GET.py": '''from typing import Any
import anyvali as av
from betterportal.handler import Handler, HandlerContext
from betterportal.registry import Operation
from betterportal.rendering import Renderer

def hello(context: HandlerContext[Any, Any, Any, Any]) -> str:
    return "Hello"

def create() -> Operation:
    page = Renderer[str]({"renderer": "bootstrap5"}, lambda value, context: "<p>Hello</p>")
    return Operation(Handler(av.string(), hello, renderers=[page]), {
        "operationId": "hello.get", "method": "GET", "title": "Hello",
        "description": "Return a greeting", "auth": {"required": True}
    })
''',
        "my_service/definition.py": f'''from betterportal.discovery import discover
from betterportal.generated_types import BpSchemaOutput, ManifestDeclarationInput
from betterportal.registry import Registry

declaration: ManifestDeclarationInput = {declaration!r}

def registry() -> Registry:
    return discover("my_service.bp_routes")

def contract() -> BpSchemaOutput:
    return registry().schema(declaration)
''',
        "my_service/app.py": '''import os
from pathlib import Path
from betterportal.asgi import create_app as asgi_app
from betterportal.bootstrap import BootstrapStateStore
from betterportal.installation import ServiceInstallation
from betterportal.service import Service
from betterportal.storage import FileStateStore
from .definition import declaration, registry

def create_app():
    directory = Path(os.environ.get("BP_STATE_DIRECTORY", ".bp-state"))
    service = Service(registry(), declaration, managed=True,
        state_store=FileStateStore(directory / "snapshot.json"))
    installation = ServiceInstallation(service,
        BootstrapStateStore(FileStateStore(directory / "bootstrap.json"), os.environ["BP_BOOTSTRAP_MASTER_KEY"]),
        os.environ["BP_CP_URL"], os.environ["BP_PUBLIC_ORIGIN"],
        settings_store=FileStateStore(directory / "settings.json"))
    return asgi_app(service, installation=installation)
''',
        "README.md": '''# Standalone BP service

Install the matching native BP packages into a Python 3.10+ environment, then run:

    python -m pip install -e .
    bp-python export --module my_service.definition:contract --output bp-contract.json
    bp-python deps sync --frozen --check

For unpublished packages, add --find-links /path/to/local/packages to pip install.
No Node or BSB installation is needed by this service.

Configure BP_CP_URL (trusted config manager), BP_PUBLIC_ORIGIN (this service's
public origin), and BP_BOOTSTRAP_MASTER_KEY through your host's protected secrets.
Generate a master key once and retain it across restarts:

    python -c "from betterportal.bootstrap import BootstrapCipher; print(BootstrapCipher.generate_key())"

BP_STATE_DIRECTORY defaults to .bp-state; mount persistent storage there in a
container. Keep the master key separate from that directory. URLs require HTTPS
except the runtime's exact loopback development exceptions.

    python -m uvicorn my_service.app:create_app --factory --host 127.0.0.1 --port 8000

Install through the existing config manager using the configured public origin.
Health at /.well-known/bp/health returns 503 until installation, manifest submission
and a valid scoped snapshot succeed. Mount hello.index / hello.get in an app;
/hello then returns JSON or Bootstrap HTML with valid user authentication.
Application ASGI lifespan owns installation, sync and shutdown cancellation.

Edit my_service/definition.py for BP identity/version and bp_routes for operations.
Keep betterportal.json and betterportal.lock.json in source control. Use native deps
commands for dependency contracts and repeat export after changing the registry.
''',
    }
    directory.mkdir(parents=True, exist_ok=False)
    for name, content in files.items():
        target = directory / name
        target.parent.mkdir(parents=True, exist_ok=True)
        with target.open("x", encoding="utf-8", newline="\n") as output:
            output.write(content)
